pywhy_graphs/algorithms/generic.py

```diff
@@ -855,3 +856,161 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
+        A set of neigbors of X.
+
+    Returns
+    -------
+    out : set
+        A set of all the neighbors of X that do not have an arrow pointing
+        back to it.
+    """
+    out = set()
+
+    for elem in Y:
+        if not (
+            G.has_edge(X, elem, G.bidirected_edge_name) or G.has_edge(elem, X, G.directed_edge_name)
+        ):
+            out.update(elem)
+
+    return out
+
+
+def _get_X_neighbors(G, X: set):
+    """Retrieve all the neigbors of X when X has more than one element.
+
+    Parameters
+    ----------
+    G : DiGraph
+        A directed graph.
+    X : Set
+
+    Returns
+    -------
+    out : set
+        A set of all the neighbors of X.
+    """
+
+    out = set()
+
+    for elem in X:
+        elem_neighbors = set(G.neighbors(elem))
+        elem_possible_neighbors = _check_back_arrow(G, elem, elem_neighbors)
+        to_remove = X.intersection(elem_possible_neighbors)
+        elem_neighbors = elem_possible_neighbors - to_remove
+
+        if len(elem_neighbors) != 0:
+            for nbh in elem_neighbors:
+                temp = (elem,)
+                temp = temp + (nbh,)
+                out.add(temp)
+    return out
+
+
+def _recursively_find_pd_paths(G, X, paths, Y):
+    """Recursively finds all the possibly directed paths for a given
+    graph.
+
+    Parameters
+    ----------
+    G : DiGraph
+        A directed graph.
+    X : Set
+        Source.
+    Y : Set
+        Destination
+
+    Returns
+    -------
+    out : set
+        A set of all the possibly directed paths.
+    """
+
+    counter = 0
+    new_paths = set()
+
+    for elem in paths:
+        cur_elem = elem[-1]
+
+        if cur_elem in Y:
+            new_paths.add(elem)
+            continue
+
+        nbr_temp = G.neighbors(cur_elem)
+        nbr_possible = _check_back_arrow(G, cur_elem, nbr_temp)
+
+        if len(nbr_possible) == 0:
+            new_paths = new_paths + (elem,)
+
+        possible_end = nbr_possible.intersection(Y)
+
+        if len(possible_end) != 0:
+            for nbr in possible_end:
+                temp_path = elem
+                temp_path = temp_path + (nbr,)
+                new_paths.add(temp_path)
+
+        remaining_nodes = nbr_possible - possible_end
+        remaining_nodes = (
+            remaining_nodes
+            - remaining_nodes.intersection(set(elem))
+            - remaining_nodes.intersection(X)
+        )
+
+        temp_set = set()
+        for nbr in remaining_nodes:
+            temp_paths = elem
+            temp_paths = temp_paths + (nbr,)
+            temp_set.add(temp_paths)
+
+        new_paths.update(_recursively_find_pd_paths(G, X, temp_set, Y))
+
+    return new_paths
+
+
+def proper_possibly_directed_path(G, X: Optional[Set] = None, Y: Optional[Set] = None):
+    """Find all the possibly directed paths in a graph.
+
+    Parameters
+    ----------
+    G : DiGraph
+        A directed graph.
+    X : Set
+        Source.
+    Y : Set
+        Destination
```

Comment:
What if both are None? I don't think None should even be allowed, but lmk if I misunderstood

Reply:
Yes you are right, neither can ever be None. Maybe I should return an empty set whenever either one of them is None.

Reply:
Why not just not have default argument?